pyblish_bumpybox/plugins/nuke/validate_nuke_write_node.py

```diff
@@ -84,10 +84,13 @@ def get_current_value(self, instance):
 
     def get_expected_value(self, instance):
 
-        expected = "[python {nuke.script_directory()}]/workspace/"
-        expected += "[python {nuke.thisNode()[\"name\"].getValue()}]/"
-        expected += "[python {os.path.splitext(os.path.basename("
-        expected += "nuke.scriptName()))[0]}]"
+        expected = (
+            "[python {nuke.script_directory()}]/workspace/[python "
```

Comment:
This is a really nasty statement, returning a bunch of python code to be interpreted later. It could be made a lot clearer by separating it out and assigning to variables with descriptive names.

I don't understand why this isn't just evaluating the python calls immediately?

Reply:
The python string is inserted into the file knob on the write node. This is only way to have a dynamic output path as far as I know in Nuke.

Do you have an alternative suggestion or solution?

Reply:
I'm not sure what you mean by dynamic output path here? Dynamic how?

The path is validated on publishing is it not? So it'll always be blocked if it doesn't match the expectations at publication time. Doesn't that make it dynamic? We're evaluating the commands and returning a real path as part of our validation. https://github.com/iwootten/pyblish-bumpybox/blob/master/pyblish_bumpybox/plugins/nuke/validate_write_file.py#L68-71

Reply:
The output path changes depending on whether the write node name is changed or the name and path of the Nuke script.
The allows us to only need to repair the path once and not every time the user want to publish.

This behaviour is the same in Maya and Houdini.

Reply:
But it leads to the same result? After a repair, but this would be expected given the write node has changed and is in line with artist behaviour. 

I don't think having python code written in paths to evaluate later is a good idea. It means artists won't understand what it represents or how to correct it. It also means any expectations will be output in log messages as code, rather than an absolute path.

Reply:
Incidentally, this is different to how the maya/houdini path validation works, as they both return absolute paths over strings.

https://github.com/iwootten/pyblish-bumpybox/blob/master/pyblish_bumpybox/plugins/maya/validate_file.py#L43

https://github.com/iwootten/pyblish-bumpybox/blob/master/pyblish_bumpybox/plugins/houdini/validate_output_path.py

Reply:
> But it leads to the same result? After a repair, but this would be expected given the write node has changed and is in line with artist behaviour.

Again in Maya and Houdini you only repair the output path once. The desired behavior here is to emulate the same, and not have to repair the write nodes settings every time you want to publish and something like the script name and the version of the script name has changed.

> I don't think having python code written in paths to evaluate later is a good idea. It means artists won't understand what it represents or how to correct it.

The question is whether the artists need to understand it? With a repair method in place people just blindly run the action anyways.
You could maybe argue that the error message for the validation could be explaining what is happening better.

> It also means any expectations will be output in log messages as code, rather than an absolute path.

If you print the evaluated file knob you'll get a full path. This is similar to how Houdini works with tokens.

> https://github.com/iwootten/pyblish-bumpybox/blob/master/pyblish_bumpybox/plugins/maya/validate_file.py#L43

This validates the location of the source textures in the Maya scene. It was nothing to do with renders, which is similarity we are drawing between Nuke and Maya.

> https://github.com/iwootten/pyblish-bumpybox/blob/master/pyblish_bumpybox/plugins/houdini/validate_output_path.py

This definitely validates the location of the output files, but you can also see that its dynamically driven by the token system in Houdini; https://github.com/iwootten/pyblish-bumpybox/blob/a784be7224da6aed81600ace1907803fd403d345/pyblish_bumpybox/plugins/houdini/validate_output_path.py#L89. This means you only have to repair the output path once.

Reply:
I'm afraid I'm going to have to disagree. Executing code that is taken from a string a user can edit is generally a bad idea in any language.

Repair type actions, as I'm led to believe are there as a necessary evil, ideally they shouldn't be used at all. The artist should by default be able to correctly update the path themselves following changes in the scene and learn to correct attributes based on an earlier failure. The problem with making an expected string include python statements is that it's difficult for anyone to correctly update them, or even know where they're pointing to on the write node. So they'll always be forced to use a repair.

Does it really matter that the repair might be run multiple times, if a write node name is modified multiple times? I'd far rather that than an odd string on a write node attribute.

Reply:
> Does it really matter that the repair might be run multiple times, if a write node name is modified multiple times?

The problem is not the changing of the write node name. I would expect this to only be set once.
The problem comes with updating the write node path when there is changed to the script name or path. This happens most often when versioning up the script name.

Reply:
BTW to repair multiple times is not a problem for me personally, but a request from the artist using it.

Reply:
> The problem comes with updating the write node path when there is changed to the script name or path. This happens most often when versioning up the script name.

Ok, so there's a change somewhere in the scene that affects the output path - be it the name of the scene or write node name.

> BTW to repair multiple times is not a problem for me personally, but a request from the artist using it.

I honestly think the impact that this has on each scene it validates is far more harmful than a subsequent repair.

Reply:
> I honestly think the impact that this has on each scene it validates is far more harmful than a subsequent repair.

This seems like we are at personal preference end pass on this.

Reply:
It's not a personal preference, executing user input is bad practice and dangerous. It's very widely discouraged.